File: 2216-delete-the-middle-node-of-a-linked-list/delete-the-middle-node-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        //Optimal by using Tortoise and Hare algorithm for finding the middle of the LL
        //with slight changes of skipping one step of slow to get it before the middle 
        //TC - O(n/2) ~O(n) and SC - O(1)
        if(head == NULL || head->next==NULL){
            return NULL;
        }
        ListNode* slow = head;
        ListNode* fast = head;
        fast = fast->next->next;
        while(fast && fast->next){
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode* delNode = slow->next;
        slow->next = slow->next->next;
        delete delNode;
        return head;

        //Brute force i can think of is findng the len of LL then len/2 then we got to that node 
        //and perform the regular deleting method of nodes woth taken care of edge cases
        //TC - O(n + n/2) ~ O(n) and SC - O(1)

        /*
        if(head==NULL || head->next==NULL){
            return NULL;
        }
        int len = 1;
        ListNode* temp = head;
        while(temp){
            len++;
            temp = temp->next;
        }
        len = (len+1)/2;
        if(len == 1){
            head->next = nullptr;
            return head;
        }
        temp = head;
        ListNode* prev = nullptr;
        while(temp){
            len--;
            if(len==0){
                prev->next = prev->next->next;
                delete temp;
                return head;
            }
            prev = temp;
            temp = temp->next;
        }
        return head;
        */
    }
};

File: 2216-delete-the-middle-node-of-a-linked-list/delete-the-middle-node-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #Optimal by using Flyod Tortoise and Hare Algorithm to find mid of LL 
        #TC - O(N) and SC - O(1)
        if not head or not head.next:
            return None

        slow = head
        fast = head
        prev = None
        while fast and fast.next:
            fast = fast.next.next
            prev = slow
            slow = slow.next
        prev.next = slow.next
        return head



        #Brute forcing by 2 pass and finding the length then dividing it by two
        #TC - O(N + N/2) and SC - O(1)
        '''
        if not head or not head.next:
            return None
        n = 0
        temp = head
        while temp:
            n += 1
            temp = temp.next
        mid = n//2
        temp = head
        prev = None
        ahead = head.next
        while mid:
            prev = temp
            temp = temp.next
            ahead = temp.next
            mid -= 1
        prev.next = ahead
        return head
        '''